app/src/main/kotlin/com/mapzen/erasermap/view/MainActivity.kt

```diff
@@ -411,11 +496,61 @@ public class MainActivity : AppCompatActivity(), ViewController,
         findViewById(R.id.route_preview).setVisibility(View.GONE)
     }
 
+
+    fun updateRoutePreview(destination: Feature?) {
+        val dest = SimpleFeature.fromFeature(destination)
+        (findViewById(R.id.by_car) as RadioButton).setOnCheckedChangeListener { compoundButton, b ->
+            if (b) {
+                val location = LocationServices.FusedLocationApi?.getLastLocation();
+                if (location is Location) {
+                    val start: DoubleArray = doubleArrayOf(location.getLatitude(), location.getLongitude())
+                    val dest: DoubleArray = doubleArrayOf(dest.getLat(), dest.getLon())
+                    getInititializedRouter().setDriving().setLocation(start).setCallback(this).setLocation(dest).fetch();
+                }
+                (findViewById(R.id.routing_circle) as ImageButton).setImageResource(R.drawable.ic_start_car_normal)
+            }
+        }
+        (findViewById(R.id.by_foot) as RadioButton).setOnCheckedChangeListener { compoundButton, b ->
+            if (b) {
+                val location = LocationServices.FusedLocationApi?.getLastLocation();
+                if (location is Location) {
+                    val start: DoubleArray = doubleArrayOf(location.getLatitude(), location.getLongitude())
+                    val dest: DoubleArray = doubleArrayOf(dest.getLat(), dest.getLon())
+                    getInititializedRouter().setWalking().setLocation(start).setLocation(dest).setCallback(this).fetch();
+                }
+                (findViewById(R.id.routing_circle) as ImageButton).setImageResource(R.drawable.ic_start_walk_normal)
+            }
+        }
+        (findViewById(R.id.by_bike) as RadioButton).setOnCheckedChangeListener { compoundButton, b ->
+            if (b) {
+                val location = LocationServices.FusedLocationApi?.getLastLocation();
+                if (location is Location) {
+                    val start: DoubleArray = doubleArrayOf(location.getLatitude(), location.getLongitude())
+                    val dest: DoubleArray = doubleArrayOf(dest.getLat(), dest.getLon())
+                    getInititializedRouter().setBiking().setLocation(start).setLocation(dest).setCallback(this).fetch()
+                }
+                (findViewById(R.id.routing_circle) as ImageButton).setImageResource(R.drawable.ic_start_bike_normal)
+            }
+        }
+    }
+
+
     override fun onBackPressed() {
+        if ((findViewById(R.id.route_preview)).getVisibility() == View.VISIBLE) {
+            mapController!!.getMap().layers().remove(path)
```

Comment:
Better to use null safe syntax here than forced dereference. `!!` will throw NPE where `?` will skip to the next line if the value is null.

            mapController?.getMap()?.layers()?.remove(path)
            mapController?.getMap()?.layers()?.remove(markers)